feat(preview): Add Pill component (#1542)
Workday/canvas-kit: modules/preview-react/pill/stories/Pill.stories.mdx

import {ClickablePillWithAvatar} from './examples/WithAvatar';
import {ClickablePillWithIcon} from './examples/WithIcon';
import {ClickablePillWithCount} from './examples/WithCount';
import {RemovablePill} from './examples/WithRemovable';
import {ListOfPills} from './examples/WithList';

<Meta title="Preview/Pill/React" component={Pill} />

# Canvas Kit Pill

Pills are static or interactive elements that allow users to input, filter, or label information.

## Installation

```sh
yarn add @workday/canvas-kit-preview-react
```

## Usage

Pills are used to visually label objects on a page for quick recognition. They’re offered as both
static, read-only and interactive elements. They allow users to input a section, filter a list or
table, or label information to help with scanning and organization.

### Basic

If only text is rendered by the Pill, it will look like a `read only` pill. This is a non
interactive element that is used to display some information.

> **_NOTE:_** By default, the Pill's width will max out at `200px` and the text will overflow into a
> tooltip. This max width can be changed by providing a `maxWidth` prop on the pill.

<ExampleCodeBlock code={Basic} />

> ## Pills with Icon, Avatar or Count

Author: ah this is awesome, updating!

All leading elements (icons or avatars) are intended to be descriptive, helping support the label.
Do not rely on the leading element to indicate the interaction behavior.

> You can define an `interactive` Pill by providing an `onClick` prop.

Reviewer: ```suggestion
You can define an interactive Pill by providing an `onClick` prop.
```

### With Count

The count appears after indicating the action is directly related to the count of that label.

<ExampleCodeBlock code={ClickablePillWithCount} />

## Pills that are removable

Pills that are removable display the `X` icon following after the label. Removable pills have a
smaller, more specific focus state and click target to be more intentional about its action and to
avoid unintended removal of a Pill.

You can define a removable pill by providing an `onDelete` click handler.

```tsx
<Pill onDelete={() => console.warn('clicked')}>
  Pink Shirts
  <Pill.IconButton />
</Pill>
```

In this case, we use a `Pill.IconButton` because the `X` becomes the focusable and clickable
element.

> **_NOTE:_** The default icon for `Pill.IconButton` is `xSmallIcon` but this can also be
> overwritten by passing an `icon` prop to `Pill.IconButton`

<ExampleCodeBlock code={RemovablePill} />

## Showing a list of Pills

Pills can often represent multiple pieces of information such as a filtered list of categories or
skills.

In order to achieve this, you can use our `HStack` component to wrap each `Pill` and space it out
accordingly.

<ExampleCodeBlock code={ListOfPills} />

> ## Components

Reviewer 2: +1 to this. 👍🏻

are acceptable

```tsx
<Pill onClick={() => console.log('clicked')}>
  <Pill.Avatar /> Regina Skeltor
</Pill>
```

If you give this component an `onDelete` it will render as a span and you can have a removable Pill
that has a smaller click target to prevent users from accidentally delete something.

```tsx
<Pill onDelete={() => console.log('handle remove')}>
  Shoes
  <Pill.IconButton />
</Pill>
```

<ArgsTable of={Pill} />

### Pill Avatar

This component renders an avatar but you can still pass it avatar props

<ArgsTable of={Pill.Avatar} />

### Pill Count

This component renders the count. The `child` element is what gets rendered as the count.

<ArgsTable of={Pill.Count} />

### Pill Icon

This component renders an `icon`. This component is used when an `onClick` is provided.

<ArgsTable of={Pill.Icon} />

### Pill IconButton

This component renders a `icon button`. This is used for `removable` Pills so that the click target
and focus area is soley the icon.

<ArgsTable of={Pill.IconButton} />

### Pill Label

This component renders a `span` that automatically handles overflow and rendering a tooltip. There's

> no need to use this component directly since the overflow is handle for you automatically.

Author: ah yep good call out, i figured id add it just in case since it is an exportable component? thoughts?

import {Meta} from '@storybook/addon-docs';
import {Specifications} from '@workday/canvas-kit-docs';

import {Pill} from '@workday/canvas-kit-preview-react/pill';

import {Basic} from './examples/Basic';
import {ClickablePillWithAvatar} from './examples/WithAvatar';
import {ClickablePillWithIcon} from './examples/WithIcon';
import {ClickablePillWithCount} from './examples/WithCount';
import {RemovablePill} from './examples/WithRemovable';
import {ListOfPills} from './examples/WithList';

<Meta title="Preview/Pill/React" component={Pill} />

# Canvas Kit Pill

Pills are static or interactive elements that allow users to input, filter, or label information.

## Installation

```sh
yarn add @workday/canvas-kit-preview-react
```

## Usage

Pills are used to visually label objects on a page for quick recognition. They’re offered as both
static (read-only) and interactive elements. They allow users to filter a list or table, or label
information to help with scanning and organization.

### Basic Example

If the Pill only renders text, it will look like a read-only Pill. This is a non-interactive element
that is used to display information.

> > **_NOTE:_** By default, the Pill's width will max out at `200px` and the text will overflow into a
> > tooltip. This max width can be changed by providing a `maxWidth` prop on the Pill.

Reviewer 2: ```suggestion
> **_NOTE:_** By default, the Pill's width will max out at `200px` and the text will be truncated with an ellipsis and render an OverflowTooltip on hover and focus. This max width can be changed by providing a `maxWidth` prop on the Pill.
```

sub components. Based on the `variant` prop this component will render different styled Pills.

#### Usage

Example of read only:

```tsx
<Pill variant="readOnly">This is a read only</Pill>
```

If you give this component a `variant='default'` prop it will render as an interactive Pill which
accepts subcomponents.

```tsx
<Pill variant="default" onClick={() => console.log('clicked')}>
  <Pill.Avatar /> Regina Skeltor
</Pill>
```

If you give this component a `variant='removable'` it will render as a span and you can have a
removable Pill that has a smaller click target to prevent users from accidentally delete something.

```tsx
<Pill variant="removable" onDelete={() => console.log('handle remove')}>
  Shoes
  <Pill.IconButton />
</Pill>
```

If you give the component an `onClick` and an `onDelete` and the `variant='default'` only `onClick`
will be called and you'll see a warning in the console suggesting to either provide an onClick or
change the variant to be `removable`.

#### Props

<ArgsTable of={Pill} />

### Pill.Avatar

This component renders an avatar. It supports all props of the `Avatar` component.

#### Usage

```tsx

> <Pill onDelete={() => console.log('handle remove')}>

Author: remove onDelete

import {Meta} from '@storybook/addon-docs';
import {Specifications} from '@workday/canvas-kit-docs';

import {Pill} from '@workday/canvas-kit-preview-react/pill';

import {Basic} from './examples/Basic';
import {ClickablePillWithAvatar} from './examples/WithAvatar';
import {ClickablePillWithIcon} from './examples/WithIcon';
import {ClickablePillWithCount} from './examples/WithCount';
import {RemovablePill} from './examples/WithRemovable';
import {ListOfPills} from './examples/WithList';

<Meta title="Preview/Pill/React" component={Pill} />

# Canvas Kit Pill

Pills are static or interactive elements that allow users to input, filter, or label information.

## Installation

```sh
yarn add @workday/canvas-kit-preview-react
```

## Usage

Pills are used to visually label objects on a page for quick recognition. They’re offered as both
static (read-only) and interactive elements. They allow users to filter a list or table, or label
information to help with scanning and organization.

### Basic Example

If the Pill has `variant='readOnly'`, it will look like a read-only Pill. This is a non-interactive
element that is used to display information.

> > **_NOTE:_** By default, the Pill's width will max out at `200px` and the text will be truncated

Reviewer 2: ```suggestion
> **_NOTE:_** `maxWidth` measures the width of the `Pill.Label` (or text) and not the width of the entire `Pill`. By default,  this `maxWidth` is set to `200px` and the text will be truncated
```

> with an ellipsis and render an OverflowTooltip on hover and focus. This max width can be changed
> by providing a `maxWidth` prop on the Pill.

<ExampleCodeBlock code={Basic} />

### Interactive Pills

All leading elements (icons or avatars) are intended to be descriptive, helping support the label.
Do not rely on the leading element to indicate the interaction behavior. By default a Pill is
considered interactive, therefore it's default variant is `default`.

```tsx
<Pill onClick={() => console.warn('clicked')}>
  <Pill.Avatar />
  Regina Skeltor
</Pill>
```

#### Avatar

The Avatar should appear before the text of a Pill. You have access to all Avatar props.

<ExampleCodeBlock code={ClickablePillWithAvatar} />

#### Icon

The `Pill.Icon` icon will default to rendering a `plusIcon`. You can change this by providing a
`icon` prop to `Pill.Icon`

<ExampleCodeBlock code={ClickablePillWithIcon} />

#### Count

> The count appears after the label. It is usually associate to the label. If you have a category, the

Reviewer 2: ```suggestion
The count appears after the label. It is usually associated with the label. If you have a category, the
```

count will dirrectly correlate to that category.

> <ExampleCodeBlock code={ClickablePillWithCount} />

Reviewer 2: ```suggestion
<ExampleCodeBlock code={WithCount} />
```

### Removable Pills

> Removable Pills display an `X` icon after the label. They have a smaller, more specific focus state
> and click target to be more intentional about their actions and to avoid unintended removal of a
> Pill.

Reviewer 2: ```suggestion
Removable `Pill`s display an `X` icon after the label. They have a smaller, more specific focus state
and click target to be more intentional about their actions and to avoid unintended removal.
```

> You can define a removable Pill by providing a `variant='removable'` prop.

Reviewer 2: ```suggestion
You can define a removable `Pill` by providing a `variant='removable'` prop.
```

```tsx
<Pill variant="removable" onDelete={() => console.warn('clicked')}>
  Pink Shirts
  <Pill.IconButton />
</Pill>
```

In this case, we use a `Pill.IconButton` because the `X` becomes the focusable and clickable
element.

The default icon for `Pill.IconButton` is `xSmallIcon` but this can also be overwritten by passing
an `icon` prop to `Pill.IconButton`

<ExampleCodeBlock code={RemovablePill} />

### List of Pills

> Pills can often represent multiple pieces of information such as a filtered list of categories or

Reviewer 2: ```suggestion
`Pill`s can often represent multiple pieces of information such as a filtered list of categories or
```

skills.

In order to achieve this, use our `HStack` component to wrap each `Pill` and space them out
accordingly.

<ExampleCodeBlock code={ListOfPills} />

## Components

### Pill

> This component is the container component for a Pill. It also provides a React Context model for its
> sub components. Based on the `variant` prop this component will render different styled Pills.

Reviewer 2: ```suggestion
This is the container component for a `Pill`. It also provides a React context model for its
subcomponents. Based on the `variant` prop this component will render different styled `Pill`s.
```

#### Usage

Example of read only:

```tsx
<Pill variant="readOnly">This is a read only</Pill>
```

> If you give this component a `variant='default'` prop it will render as an interactive Pill which
> accepts subcomponents.
> 
> ```tsx
> <Pill variant="default" onClick={() => console.log('clicked')}>
>   <Pill.Avatar /> Regina Skeltor
> </Pill>
> ```

Reviewer 2: Nit: move this above the `readOnly` example.

```suggestion
By default, a `Pill` render an interactive element that accepts subcomponents.

```tsx
<Pill onClick={() => console.log('clicked')}>
  <Pill.Avatar /> Regina Skeltor
</Pill>
```
```

> If you give this component a `variant='removable'` it will render as a span and you can have a
> removable Pill that has a smaller click target to prevent users from accidentally delete something.

Reviewer 2: ```suggestion
If you set the `Pill` `variant` to `removable`, it will render a `<span>` element. You can then provide a `Pill.IconButton` that acts as the focus target. This creates a smaller, more intentional click target that prevents users from accidentally deleting an item.
```

```tsx
<Pill variant="removable" onDelete={() => console.log('handle remove')}>
  Shoes
  <Pill.IconButton />
</Pill>
```

> If you give the component an `onClick` and an `onDelete` and the `variant='default'` only `onClick`
> will be called and you'll see a warning in the console suggesting to either provide an onClick or
> change the variant to be `removable`.

Reviewer 2: ```suggestion
In the example above, `onDelete` will only fire when the `Pill.IconButton` has been clicked.
```

#### Props

<ArgsTable of={Pill} />

### Pill.Avatar

This component renders an avatar. It supports all props of the `Avatar` component.

#### Usage

```tsx
<Pill onDelete={() => console.log('handle remove')}>
  <Pill.Avatar url={avatarUrl} />
  Regina Skeltor
  <Pill.IconButton />
</Pill>
```

#### Props

<ArgsTable of={Pill.Avatar} />

### Pill.Count

This component renders its `children` as the count.

#### Usage

```tsx
<Pill onClick={() => console.warn('clicked')}>
  Shoes
  <Pill.Count>30</Pill.Count>
</Pill>
```

#### Props

<ArgsTable of={Pill.Count} />

### Pill.Icon

> This component renders an `icon`. This component should only be used when `variant='default'`. By
> default this component render a `plusIcon` but it can be overriden.

Reviewer 2: ```suggestion
This component renders an `icon`. It not be used with the `default` styling – not `readOnly` or `removable` variants. By default it renders a `plusIcon` but it can be overridden by providing an icon to the `icon` prop.
```

#### Usage

```tsx
<Pill onClick={() => console.warn('clicked')}>
  <Pill.Icon />
  <Pill.Label>Regina Skeltor</Pill.Label>
</Pill>
```

#### Props

<ArgsTable of={Pill.Icon} />

### Pill.IconButton

> This component renders an icon button. This is used for removable Pills where the click target and
> focus area is solely the icon. By default, this component renders a `xSmallIcon` but can be
> overriden.

Reviewer 2: ```suggestion
This component renders a custom icon button. It is only intended to be used with the `removable` variant. By default, it renders a `xSmallIcon` but can be overridden by providing an icon to the `icon` prop.
```
